Introduced point tools for statistical analysis

Idclip opened the pull request:
Disclaimer: This work was done as a requirement for a subsequent PR. While I'm fairly happy with the API, there may be better implementations which I haven't considered. I've been using this alongside a much heavier (slower) tool so have not been focusing on squeezing every last drop of performance here (though I think it's okay for a first attempt).

This is a pre-requisite to the surfacing PR. It follows a typical VDB style API with three main methods:
 - `points::evalMinMax`
 - `points::evalAverage`
 - `points::accumulate`

Main things to note:
 - I wanted the ValueTypes to adhere to strict rules so that we don't run into the same problem we did with matrix grids evalMinMax. These rules are set out in the function descriptions and are as minimal as possible. For example, evalMinMax does not require numeric limits support or zero val initialization etc.
 - These methods differ significantly from existing voxel methods due to point filtering and need to handle cases such as no valid points for a given filter.
 - These methods support the construction of equivalent tiled trees with localised results.

Signed-off-by: Nick Avramoussis <4256455+Idclip@users.noreply.github.com>

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+            for (auto leaf = range.begin(); leaf; ++leaf) {
+                AttributeHandle<ValueT, CodecT> handle(leaf->constAttributeArray(idx));
+                if (handle.size() == 0) continue;
+                if (std::is_same<FilterT, NullFilter>::value) {
```
@danrbailey we really need to talk about point filters. I would love to be able to use what I have here i.e. if a `NullFilter` is used, iterate directly on the array. It is so much faster than using the `IndexFilter` API. Additionally, we never actually conform to the `ActiveFilter`/`BinaryFilter` tools. Functions seem to only iteration over ValueOn data, but surely this is wrong? Tools should default to a `FilterT=ActiveFilter` and should  use `leaf->beginIndexAll()`. Otherwise these filter become redundant. In so doing, the `NullFilter` option provides direct access to doing ` for (size_t i = 1; i < size; ++i)` which is much nicer.

Idclip replied:
@danrbailey be good to chat about this as I feel like it's a considerable limitation in the points API. But I don't think it should stop this going on.

Idclip replied:
Done! I swear when i originally implemented this there was some issue with filter.state() which wasn't giving me the desired results but that's so long ago I might be misremembering. Works great for now and will keep an eye on it.

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+///   deterministic.
+/// @note  The value type of the min/max calculations must match the value type
+///   of the attribute. For vectors and matrices, this results in per component
+///   comparisons. See evalExtents for magnitudes or more custom control.
```
By default, `evalMinMax` works per component. There is an internal function `evalExtents` which I have not exposed, but see the unit tests for an example: https://github.com/AcademySoftwareFoundation/openvdb/blob/eeddd9b0b3eb4028f3616825406765b4e4109a59/openvdb/openvdb/unittest/TestPointStatistics.cc#L739

It allows for custom results to be produces for things like magnitudes

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+    typename ResultT = ValueT,
+    typename CodecT = UnknownCodec,
+    typename FilterT = NullFilter,
+    typename PointDataTreeT = points::PointDataTree>
```
I don't like that you have to provide it when 99.9% of the time it's the default configuration. i.e, if it's not last and not default, I have to provide it a redundant value just to change, say, the `CodecT `, even though it's the default and deducible. 

> given that the argument on which it's based is mandatory.

What do you mean by this? The arg is deducible so the fact that it's defaulted doesn't matter... should you call this method with something else the compiler will generate a different signature. The order in which the compiler decides what the value of function template arguments are, is roughly:

1. the template argument is explicitly provided
2. the compiler deduces it from the context
3. uses the default value. 

i.e. a default template argument never overrides a given/deduced one.

Idclip replied:
Removed default argument for PointDataTreeT

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+        ValueT& max,
+        const FilterT& filter = NullFilter(),
+        typename PointDataTreeT::template ValueConverter<ValueT>::Type* minTree = nullptr,
+        typename PointDataTreeT::template ValueConverter<ValueT>::Type* maxTree = nullptr);
```
Can you elaborate on why you think this is "clunky"? These are optional arguments e.g:
```
evalMinMax<int32_t>(tree, min, max, NullFilter());
...
Int32Tree minTree;
evalMinMax<int32_t>(tree, min, max, NullFilter(), &minTree);
```
I don't want to pollute the public API where possible.

> You could compute the former directly from the later pretty inexpensively?

Yes, but if you don't want these trees you shouldn't have to pay the penalty (more memory and slightly slower) to create them. In other words, they should not be constructed if not provided and we can utilise a full reduction rather than an extra post process.

Idclip replied:
Added basic function singatures!

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+    typename CodecT = UnknownCodec,
+    typename FilterT = NullFilter,
+    typename PointDataTreeT = points::PointDataTree>
+inline bool evalMinMax(const PointDataTreeT& points,
```
I guess I originally wrote this to match `Grid::evalMinMax()` but there's no reason a `evalMin` and `evalMax` can't exist too - I needed both though. I think they could be added later.

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+///   type ResultT to be provided which is used to compute the final result
+///   as well as for intermediate arithmetic. This allows for higher precision
+///   results if overflow or underflow is a concern, as well as computing
+///   floating point results from integer types. This method will throw only if
```
It's impossible to assume all use cases - I personally would find it strange that a reduction performed at a different (higher) precision than my requested `ValueType`. The same goes for integers, it may be that a point set has been configured such that its values and counts are divisible. These are perhaps edge cases - I wouldn't be completely opposed to removing this option but I'd like to hear if anyone else has thoughts.

Idclip replied:
Note that this also supports a different destination type for custom types

Idclip replied:
As discussed, I will be changing `average` to work at double precision for floats and integers and to always return a double. `accumulate` will work at the higher corresponding type precision and return that type.

Idclip replied:
I've added new type traits classes to handle this. The thing is that I still wanted the `ResultT` type to be templated so that the function is extensible to future/custom types. At the moment both `average` and `accumulate` will change their precision based on the input result type:


```
template <typename ValueT,
    typename ResultT = typename ConvertType<ValueT, double>::Type,
     ....>
inline bool evalAverage(const PointDataTreeT& points,
        ResultT& average,
        ...);

double d;
evalAverage</*attr type*/float>(points, d); // runs at double precision

float f; 
evalAverage</*attr type*/float>(points, f); // runs at float precision
```

The alternative would be to do:

```
template <typename ValueT,
     ....>
inline bool evalAverage(const PointDataTreeT& points,
         typename ConvertType<ValueT, double>::Type& average,
        ...);

double d;
evalAverage</*attr type*/float>(points, d); // runs at double precision

// float f; 
// evalAverage</*attr type*/float>(points, f); // DOESN'T COMPILE
```

I feel this is pretty limiting. @danrbailey thoughts?

Idclip replied:
This is now resolved for both average and total methods

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+/// @param averageTree  if provided, builds a tiled tree of localised avg results
+/// @return true if average has been set, false otherwise. Can be false if
+///   no points were processed or if the tree was empty.
+/// @warning if "P" is provided as the attributes, it is evaluated in voxel space
```
I agree, we need unique methods/algorithms for position - they don't necessarily need to take a grid though, it could just compute it in index space. But I don't think we should explicitly error here unless we see these methods supporting index space P in the future (in which case we error and add this support later).

jmlait replied:
Average of P is an obvious thing fro computing barycenters, so I'm wondering if there is a more explicity way we could catch this....

Like maybe make it *undefined* and eventually this could switch automatically to evaluatePosition if a voxel-relative attribute is selected?

Idclip replied:
The result for all these methods, if called with 'P', is now described as undefined

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+
+/// @brief  Vector component-wise op to evaluate the min/max of
+///   vector components and return the result as a vector of
+///   equal size and precision
```
I don't have a use case for any vec/mat min/maxing at the moment :)

jmlait replied:
Compute bounding box of rest attribute?

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+#include "../openvdb.h"
+#include "../Types.h"
+#include "../math/Math.h"
+#include "../tree/LeafManager.h"
```
Done

richhones commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,611 @@
+        }
+    }
+
+    average = ResultT(average / static_cast<DivisorT>(count));
```
The way this works if I have read it correctly, it calcs the average per leaf, stores that the in values vector and then divides by the number of leaves (i.e. non null elements in values) right? If so I dont think that works for averaging. I.e. if a leaf had one point of 6 with attribute that was 1, the rest 0 (i.e. values[idx] = 1/6) and then another leaf had one point that was 1 and the rest 0, from 2 points (i.e. values[otheridx] = 1/2), then this assumes (1/6 + 1/2)/2 == 2/8 but that isnt true. (1/6 + 1/2)/2 = 1/3. I assume this isnt picked up as the tests all use the same number of active points in each leaf?

Idclip replied:
This was a silly mistake thanks for picking this up @richhones. Fixed

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,609 @@
+///   point attributes. The ValueType of the attribute must be copy
+///   constructible and support the same type addition operator. This method
+///   allows for a different type ResultT to be provided which is used to
+///   compute the final result as well as for intermediate arithmetic. This
```
Done

Idclip commented:
> Anyway, I have lots more questions, but perhaps we can schedule a quick session to talk through some of this to try and get this ready in time for v9?

Okay. Fundamentally I think we need a better way to speed up this process. I'm happy to iterate on the API but if we accept that it might take half a year for a review (which I do understand) then we should accept merging stable solutions and changing them in the future.

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,611 @@
+///   supports arbitrary point filters.
+///
+
+#ifndef OPENVEB_POINTS_STATISTICS_HAS_BEEN_INCLUDED
```
Done

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,650 @@
+template <typename ValueT,
+    typename CodecT = UnknownCodec,
+    typename FilterT = NullFilter,
+    typename PointDataTreeT = points::PointDataTree>
```
Left `ValueT` where it is but removed the default arg for `PointDataTreeT`

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,650 @@
+    typename CodecT = UnknownCodec,
+    typename FilterT = NullFilter,
+    typename PointDataTreeT = points::PointDataTree>
+inline bool evalMinMax(const PointDataTreeT& points,
```
So I've removed the inline on these and the new methods but this got me wondering, we basically have:

```
void basic_signature();
void impl_signature();


void basic_signature()
{
    impl_signature();
}

void impl_signature()
{
    // stuff
}
```

Does it not make sense to allow the trivial wrappers to remain inlined?

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,650 @@
+    const size_t idx = manager.leaf(0).attributeSet().find(attribute);
+    if (idx == AttributeSet::INVALID_POS) return false;
+
+    std::vector<std::unique_ptr<Sample>> values;
```
As discussed leaving as is for now

Idclip commented on openvdb/openvdb/Types.h:
```diff
@@ -272,6 +272,85 @@ struct ValueTraits<T, false>
+template<typename T> struct PromoteType<math::Quat<T>> : public PromoteContainerType<T, math::Quat> {};
+template<typename T> struct PromoteType<math::Mat3<T>> : public PromoteContainerType<T, math::Mat3> {};
+template<typename T> struct PromoteType<math::Mat4<T>> : public PromoteContainerType<T, math::Mat4> {};
+/// @}
```
Added a unit test

Idclip commented on openvdb/openvdb/points/PointStatistics.h:
```diff
@@ -0,0 +1,650 @@
+#include "openvdb/openvdb.h"
+#include "openvdb/Types.h"
+#include "openvdb/math/Math.h"
+#include "openvdb/tree/LeafManager.h"
```
Done

Idclip commented:
@danrbailey I've made the following changes:

     - Introduced basic variants of evalMinMax, evalAverage and accumulate
     - Changed is_same<FilterT, NullFilter> to check filter.state()
     - Removed inline keyword from methods
     - Removed default argument from some templates
     - Added unit test for Convert/PromoteType
     - Fixed include guard spelling
     - Changed include brackets

danrbailey reviewed (commented):
Thanks for making all these changes Nick, this looks great. 

I have a question about the function signature for `evalMinMax()`. I would prefer if we use `math::MinMax` from `math/Stats.h` as it makes for a slightly clearer and more consistent API in my opinion. Is there a good reason not to use that? 

Also, that object has default values of `min=std::numeric_limits<>::max()` and `max=std::numeric_limits<>::lowest()` which makes it possible to tell when the PointDataGrid is empty or the filter evaluates to nothing. Using zeroVal() for both means that it's not possible to distinguish from that and a set of attributes that are all zero-valued. Likewise for `evalAverage()` and `accumulate()`, in my opinion it would be better to return a limit value like `std::numeric_limits<>::max()` if nothing matches instead of zero (which is arguably a much more likely attribute value). 

Other than that, I'm happy with all the other changes here.

Idclip commented:
> I have a question about the function signature for evalMinMax(). I would prefer if we use math::MinMax from math/Stats.h as it makes for a slightly clearer and more consistent API in my opinion. Is there a good reason not to use that?

The main reason was that this new method does **not** require that a `ValueType` define its own specialization for `std::numeric_limits `. It also does not need to support `<<` (required by the `MinMax` struct). Both of these are currently unnecessary.  

> Also, that object has default values of min=std::numeric_limits<>::max() and max=std::numeric_limits<>::lowest() which makes it possible to tell when the PointDataGrid is empty or the filter evaluates to nothing. 

It doesn't really. These are just as valid values as zero. Perhaps they are less likely but unless the standard or VDB itself starts to say they have some special meaning they're just as valid an input. 

> Using zeroVal() for both means that it's not possible to distinguish from that and a set of attributes that are all zero-valued. 

This is exactly what the function that returns true/false was for in the first place!


Ultimately I wanted to restrict what `ValueType` has to adhere to to what is absolutely necessary. I think I might be able to add this support to these wrapper methods without restricting the requirement on the main ones (though I won't be using `evalMinMax::add`) so I'll give this a shot.

danrbailey commented:
> > I have a question about the function signature for evalMinMax(). I would prefer if we use math::MinMax from math/Stats.h as it makes for a slightly clearer and more consistent API in my opinion. Is there a good reason not to use that?
> 
> The main reason was that this new method does **not** require that a `ValueType` define its own specialization for `std::numeric_limits `. It also does not need to support `<<` (required by the `MinMax` struct). Both of these are currently unnecessary.

Makes sense. Certainly the `<<` requirement is a bit unnecessary, I would be in favour of removing that. Are there are any types that are defined by default where that is not the case? I'm happy to accept the numeric_limits requirement if so, otherwise I'm fine with your pair solution here. Just would be nice to be consistent with the new `tools::minMaxValues()`, etc.

> > Also, that object has default values of min=std::numeric_limits<>::max() and max=std::numeric_limits<>::lowest() which makes it possible to tell when the PointDataGrid is empty or the filter evaluates to nothing.
> 
> It doesn't really. These are just as valid values as zero. Perhaps they are less likely but unless the standard or VDB itself starts to say they have some special meaning they're just as valid an input.

Actually, that's not true, the range is (**max**, **min**). Regardless of the value, the result can never be (**max**, **min**) even if one value matched as max is never less than min.

> > Using zeroVal() for both means that it's not possible to distinguish from that and a set of attributes that are all zero-valued.
> 
> This is exactly what the function that returns true/false was for in the first place!
> 
> Ultimately I wanted to restrict what `ValueType` has to adhere to to what is absolutely necessary. I think I might be able to add this support to these wrapper methods without restricting the requirement on the main ones (though I won't be using `evalMinMax::add`) so I'll give this a shot.

I do understand what you're saying here, I just think given that it's less likely, it's a more sensible default.

Idclip commented:
> Actually, that's not true, the range is (max, min). Regardless of the value, the result can never be (max, min) even if one value matched as max is never less than min.

Ah yes this is correct, I was more talking about returning this value from `average` or `accumulate`

Idclip commented:
> Makes sense. Certainly the << requirement is a bit unnecessary, I would be in favour of removing that. Are there are any types that are defined by default where that is not the case? I'm happy to accept the numeric_limits requirement if so, otherwise I'm fine with your pair solution here. Just would be nice to be consistent with the new tools::minMaxValues(), etc.

@danrbailey  this is currently broken for VDB Math types. `std::numeric_limits<Vec3f>::max()` currently returns `[0,0,0]` as it's not implemented. Additionally `tree.evalMinMax()` returns `zeroVal<T>()` for empty trees.

danrbailey commented:
> > Makes sense. Certainly the << requirement is a bit unnecessary, I would be in favour of removing that. Are there are any types that are defined by default where that is not the case? I'm happy to accept the numeric_limits requirement if so, otherwise I'm fine with your pair solution here. Just would be nice to be consistent with the new tools::minMaxValues(), etc.
> 
> @danrbailey this is currently broken for VDB Math types. `std::numeric_limits<Vec3f>::max()`, for example, currently returns `[0,0,0]` as it's not implemented. Additionally `tree.evalMinMax()` returns `zeroVal<T>()` for empty trees.

Fair enough. We should really address the fact that not all our custom types have numeric_limits specializations, but it doesn't feel like a deal-breaker to resolve this for this PR so I'll approve as-is. Thank you for your patience and for adding the simpler function signatures.

Idclip commented:
Thanks @danrbailey. Did you see my question about inlining?

> 

> So I've removed the inline on these and the new methods but this got me wondering, we basically have:
```C++
void basic_signature();
void impl_signature();

void basic_signature() // inline?
{
    impl_signature();
}
void impl_signature()
{
    // stuff
}
```
> Does it not make sense to allow the trivial wrappers to remain inlined?

danrbailey commented:
> Thanks @danrbailey. Did you see my question about inlining?
> 
> > 
> 
> > So I've removed the inline on these and the new methods but this got me wondering, we basically have:
> 
> ```c++
> void basic_signature();
> void impl_signature();
> 
> void basic_signature() // inline?
> {
>     impl_signature();
> }
> void impl_signature()
> {
>     // stuff
> }
> ```
> 
> > Does it not make sense to allow the trivial wrappers to remain inlined?

Yes, sorry I did see your comment. The inline keyword here has nothing to do with inlining, the compiler will ignore your hint and makes it's own determination as to whether inlining is worthwhile or not. Typically, the inline keyword is actually for ODR, however this is a templated method so ODR does not apply. As a result, the _**only**_ thing the inline keyword does on a templated non-member function is to disable explicit template instantiation. As a result, the rule of thumb is to never use the inline keywords for templated non-member functions.
